README.md
yandex-speller
==============
[![NPM version](https://img.shields.io/npm/v/yandex-speller.svg)](https://www.npmjs.com/package/yandex-speller)
[![NPM Downloads](https://img.shields.io/npm/dm/yandex-speller.svg?style=flat)](https://www.npmjs.org/package/yandex-speller)
[![Coverage Status](https://img.shields.io/coveralls/hcodes/yandex-speller.svg)](https://coveralls.io/r/hcodes/yandex-speller)
[![Dependency Status](https://img.shields.io/david/hcodes/yandex-speller.svg)](https://david-dm.org/hcodes/yandex-speller)

[Yandex.Speller API](https://tech.yandex.ru/speller/doc/dg/concepts/api-overview-docpage/) for Node.js.

## Installation
`npm install yandex-speller`

## API

### [.checkText](./lib/yandex-speller.js#L14)(text, callback, settings)
### [.checkTexts](./lib/yandex-speller.js#L39)(texts, callback, settings)

## [Restrictions API Yandex.Speller](http://legal.yandex.ru/speller_api/)

## [License](./LICENSE.md)
MIT License


CHANGELOG.md
# Changelog

## 5.0.0
- Drop support Node.js < 16.
- Fix timeout event.
- Update dev deps in package.json.

## 4.1.0
Add error codes for export #6.

## 4.0.0
Revert switch to new Yandex Speller API.

## 3.0.0
Switch to new Yandex Speller API.
Removed unsupported API options:
- IGNORE_UPPERCASE
- IGNORE_DIGITS
- FIND_REPEAT_WORDS
- IGNORE_LATIN:
- NO_SUGGEST
- FLAG_LATIN
- BY_WORDS
- IGNORE_CAPITALIZATION
- IGNORE_ROMAN_NUMERALS
- IGNORE_CAPITALIZATION

## 2.1.0
Repeating server request in case of internet connection issues #1 @Avetome

## 2.0.1
Fixed `use strict` mode.

## 2.0.0
- A support for old Node version is dropped
- Removed module “request” from deps
- Fixed codestyle (jshint → eslint)

## 1.0.0
Initial public release.


test/test.js
'use strict';

const yaspeller = require('../lib/yandex-speller');
const assert = require('chai').assert;
const fs = require('fs');
const getFile = name => fs.readFileSync(name).toString('utf-8');

describe('checkText()', function() {
    this.timeout(10000);

    it('ignoreUppercase on', function(done) {
        const text = getFile('./test/texts/ignore_uppercase.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'ru',
            options: {ignoreUppercase: true}
        });
    });

    it('ignoreUppercase on, html', function(done) {
        const text = getFile('./test/texts/ignore_uppercase_html.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'ru,en',
            format: 'html',
            options: {ignoreUppercase: true}
        });
    });

    it('ignoreUppercase off', function(done) {
        var text = getFile('./test/texts/ignore_uppercase.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {lang: 'ru'});
    });

    it('ignoreDigits on', function(done) {
        const text = getFile('./test/texts/ignore_digits.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {lang: 'ru', options: {ignoreDigits: true}});
    });

    it('ignoreDigits off', function(done) {
        const text = getFile('./test/texts/ignore_digits.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {lang: 'ru', format: 'plain'});
    });

    it('ignoreLatin on', function(done) {
        const text = getFile('./test/texts/ignore_latin.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: ['en', 'ru'],
            format: 'plain',
            options: {ignoreLatin: true}
        });
    });

    it('ignoreLatin off', function(done) {
        const text = getFile('./test/texts/ignore_latin.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        });
    });

    it('ignoreUrls on', function(done) {
        const text = getFile('./test/texts/ignore_urls.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'en,ru',
            format: 'plain',
            options: {ignoreUrls: true}
        });
    });

    it('ignoreUrls off', function(done) {
        const text = getFile('./test/texts/ignore_urls.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 2);
            done();
        });
    });

    it('ignoreCapitalization on', function(done) {
        const text = getFile('./test/texts/ignore_capitalization.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'ru',
            format: 'plain',
            options: {ignoreCapitalization: true}
        });
    });

    it('ignoreCapitalization off', function(done) {
        const text = getFile('./test/texts/ignore_capitalization.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {lang: 'ru'});
    });

    it('findRepeatWords on', function(done) {
        const text = getFile('./test/texts/find_repeat_words.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {
            lang: 'ru',
            options: {findRepeatWords: true}
        });
    });

    it('findRepeatWords off', function(done) {
        const text = getFile('./test/texts/find_repeat_words.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'ru',
            format: 'plain'
        });
    });

    it('ignoreRomanNumerals on', function(done) {
        const text = getFile('./test/texts/ignore_roman_numerals.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {
            lang: 'en,ru',
            format: 'plain',
            options: {ignoreRomanNumerals: true}
        });
    });

    it('ignoreRomanNumerals off', function(done) {
        const text = getFile('./test/texts/ignore_roman_numerals.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {lang: 'en,ru', format: 'plain'});
    });

    it('flagLatin on', function(done) {
        const text = getFile('./test/texts/flag_latin.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 1);
            done();
        }, {
            lang: 'ru',
            format: 'plain',
            options: {flagLatin: true}
        });
    });

    it('flagLatin off', function(done) {
        const text = getFile('./test/texts/flag_latin.txt');
        yaspeller.checkText(text, function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 0);
            done();
        }, {lang: 'ru', format: 'plain'});
    });
});

describe('checkTexts()', function() {
    it('ignoreUppercase on, html', function(done) {
        const text = getFile('./test/texts/ignore_uppercase_html.txt');
        yaspeller.checkTexts([text, text], function(err, data) {
            assert.equal(err, null);
            assert.equal(data.length, 2);
            assert.equal(data[0].length, 2);
            assert.equal(data[1].length, 2);
            done();
        }, {
            lang: 'ru,en',
            options: {ignoreUppercase: true}
        });
    });
});


lib/post.js
'use strict';

const querystring = require('querystring');
const http = require('http');

const ETIMEDOUT = 'Request timed out.';

function post(params, callback) {
    const form = querystring.stringify(params.form);
    const options = {
        host: params.host,
        path: params.path,
        port: 80,
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': Buffer.byteLength(form)
        }
    };

    let req = http.request(options, function(res) {
        let textBuffer = '';

        res.setEncoding('utf8');

        res.on('data', function(chunk) {
            textBuffer += chunk;
        });

        res.on('end', function() {
            let json;
            try {
                json = JSON.parse(textBuffer);
            } catch (e) {
                callback(
                    'Yandex.Speller API:\n-------------------\n' +
            'Status code: ' + res.statusCode + '\n' +
            'Response: ' + textBuffer + '\n' +
            'JSON.parse(Response): ' + e + '\n'
                );

                return;
            }

            callback(null, res, json);
        });
    });

    req.on('error', function(e) {
        callback(e);
    });

    req.on('timeout', function() {
        req.destroy(Error(ETIMEDOUT));
    });

    req.write(form);
    req.end();
}

function postRetryingDecorator(params, callback) {
    let tryNumber = 1;
    let callbackOrigin = callback;
    let paramsOrigin = params;
    let requestLimit = params.requestLimit || 2;
    let timeout = params.timeout || 500;

    let callbackMock = function(err) {
        if (err && err.message == ETIMEDOUT && tryNumber <= requestLimit) {
            tryNumber++;
            setTimeout(function() { post(paramsOrigin, callbackMock); }, timeout);
            return;
        }

        callbackOrigin.apply(this, arguments);
    };

    post(params, callbackMock);
}

module.exports = postRetryingDecorator;


lib/yandex-speller.js
'use strict';

const post = require('./post');
const {
    errors,
    ERROR_UNKNOWN_WORD,
    ERROR_REPEATED_WORD,
    ERROR_CAPITALIZATION,
    ERROR_TOO_MANY_ERRORS,
} = require('./errors');

const YASPELLER_HOST = 'speller.yandex.net';
const YASPELLER_PATH = '/services/spellservice.json/';

const DEFAULT_FORMAT = 'plain';
const DEFAULT_LANG = 'en,ru';
const DEFAULT_REQUEST_LIMIT = 2;
const DEFAULT_TIMEOUT = 500;

/**
 * Check text for typos.
 *
 * @param {string} text
 * @param {Function} callback
 * @param {Settings} settings
 * @see {@link https://tech.yandex.ru/speller/doc/dg/reference/checkText-docpage/}
 */
function checkText(text, callback, settings) {
    const form = prepareSettings(settings);
    form.text = text;

    post({
        host: YASPELLER_HOST,
        path: YASPELLER_PATH + 'checkText',
        form: form,
    }, function(error, response, body) {
        if (error) {
            callback(error, null);
        } else {
            if (response.statusCode === 200) {
                callback(null, body);
            } else {
                callback(
                    Error('Yandex.Speller API returns status code is ' +
              response.statusCode, null)
                );
            }
        }
    });
}

/**
 * Check text for typos.
 *
 * @param {string[]} texts
 * @param {Function} callback
 * @param {Settings} settings
 * @see {@link https://tech.yandex.ru/speller/doc/dg/reference/checkTexts-docpage/}
 */
function checkTexts(texts, callback, settings) {
    const form = prepareSettings(settings);
    form.text = texts;

    post({
        host: YASPELLER_HOST,
        path: YASPELLER_PATH + 'checkTexts',
        form: form,
    }, function(error, response, body) {
        if (error) {
            callback(error, null);
        } else {
            if (response.statusCode === 200) {
                callback(null, body);
            } else {
                callback(
                    Error('Yandex.Speller API returns status code is ' +
              response.statusCode, null)
                );
            }
        }
    });
}

function prepareOptions(options) {
    let result = 0;
    // https://tech.yandex.ru/speller/doc/dg/reference/speller-options-docpage/
    const standartOptions = {
        IGNORE_UPPERCASE: 1,
        IGNORE_DIGITS: 2,
        IGNORE_URLS: 4,
        FIND_REPEAT_WORDS: 8,
        IGNORE_LATIN: 16,
        NO_SUGGEST: 32,
        FLAG_LATIN: 128,
        BY_WORDS: 256,
        IGNORE_CAPITALIZATION: 512,
        IGNORE_ROMAN_NUMERALS: 2048,
    };

    Object.keys(options || {}).forEach(function(key) {
        const upperCaseKey = key.replace(/([A-Z])/g, '_$1').toUpperCase();
        if (standartOptions[upperCaseKey] && options[key]) {
            result |= standartOptions[upperCaseKey];
        }
    });

    return result;
}

function prepareSettings(settings) {
    settings = settings || {};

    return {
        format: settings.format || DEFAULT_FORMAT,
        lang: settings.lang || DEFAULT_LANG,
        options: prepareOptions(settings.options),
        requestLimit: settings.requestLimit || DEFAULT_REQUEST_LIMIT,
        timeout: settings.timeout || DEFAULT_TIMEOUT,
    };
}

/**
 * @typedef {Object} Settings
 * @param {string} [format] Text format: plain or html.
 * @param {string|Array} [lang] Language: en, ru or uk.
 * @param {number} [requestLimit] Request repeat count in case internet connection issues.
 * @param {number} [timeout] Timeout between request repeats in milliseconds.
 * @param {Object} [options]
 * @param {boolean} [options.ignoreUppercase] Ignore words written in capital letters.
 * @param {boolean} [options.ignoreDigits] Ignore words with numbers, such as "avp17h4534".
 * @param {boolean} [options.ignoreUrls] Ignore Internet addresses, email addresses and filenames.
 * @param {boolean} [options.findRepeatWords] Highlight repetitions of words, consecutive. For example, "I flew to to to Cyprus".
 * @param {boolean} [options.ignoreLatin] Ignore words, written in Latin, for example, "madrid".
 * @param {boolean} [options.noSuggest] Just check the text, without giving options to replace.
 * @param {boolean} [options.flagLatin] Celebrate words, written in Latin, as erroneous.
 * @param {boolean} [options.byWords] Do not use a dictionary environment (context) during the scan. This is useful in cases where the service is transmitted to the input of a list of individual words.
 * @param {boolean} [options.ignoreCapitalization] Ignore the incorrect use of UPPERCASE / lowercase letters, for example, in the word "moscow".
 * @param {boolean} [options.ignoreRomanNumerals] Ignore Roman numerals ("I, II, III, ...").
 */

module.exports = {
    checkText,
    checkTexts,

    defaultFormat: DEFAULT_FORMAT,
    defaultLang: DEFAULT_LANG,

    errors,
    errorCodes: errors, // Deprecated
    
    ERROR_UNKNOWN_WORD,
    ERROR_REPEATED_WORD,
    ERROR_CAPITALIZATION,
    ERROR_TOO_MANY_ERRORS,
    
    supportedFormats: ['plain', 'html']
};
